core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1969,34 +1951,55 @@ class KafkaApis(val requestChannel: RequestChannel,
       throw new ClusterAuthorizationException(s"Request $request is not authorized.")
   }
 
-  private def sendResponseMaybeThrottle(request: RequestChannel.Request, createResponse: Int => AbstractResponse) {
-    sendResponseMaybeThrottle(request, request.header.clientId, { requestThrottleMs =>
-      sendResponse(request, createResponse(requestThrottleMs))
-    })
+  private def sendResponseMaybeThrottle(request: RequestChannel.Request, createResponse: Int => AbstractResponse): Unit = {
+    quotas.request.maybeRecordAndThrottle(request,
+      throttleTimeMs => sendResponse(request, Some(createResponse(throttleTimeMs))))
   }
 
-  private def sendResponseMaybeThrottle(request: RequestChannel.Request, clientId: String, sendResponseCallback: Int => Unit) {
-
-    if (request.apiRemoteCompleteTimeNanos == -1) {
-      // When this callback is triggered, the remote API call has completed
-      request.apiRemoteCompleteTimeNanos = time.nanoseconds
+  private def sendErrorResponseMaybeThrottle(request: RequestChannel.Request, error: Throwable) {
```

Comment:
We have a few `sendErrorResponse` inner methods, could they be simplified if they used the new `sendError*` methods?